core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala

```diff
@@ -0,0 +1,846 @@
+   * @param groupId the groupId of the group we are removing
+   */
+  def removeGroup(groupId: String) {
+    if (!groupsCache.contains(groupId))
+      throw new IllegalArgumentException("Cannot remove non-existing group")
+    groupsCache.remove(groupId)
+  }
+
+  def storeGroup(group: GroupMetadata,
+                 groupAssignment: Map[String, Array[Byte]]): Short = {
+    // construct the message to append
+    val message = new Message(
+      key = GroupMetadataManager.groupMetadataKey(group.groupId),
+      bytes = GroupMetadataManager.groupMetadataValue(group, groupAssignment)
+    )
+
+    val groupMetadataPartition = TopicAndPartition(GroupCoordinator.GroupMetadataTopicName, partitionFor(group.groupId))
+
+    val groupMetadataMessageSet = Map(groupMetadataPartition ->
+      new ByteBufferMessageSet(config.offsetsTopicCompressionCodec, message))
+
+    val responseCode = Errors.NONE.code
+    val responseLatch = new CountDownLatch(1)
+
+    // set the callback function to insert offsets into cache after log append completed
+    def putCacheCallback(responseStatus: Map[TopicAndPartition, ProducerResponseStatus]) {
+      // the append response should only contain the topics partition
+      if (responseStatus.size != 1 || ! responseStatus.contains(groupMetadataPartition))
+        throw new IllegalStateException("Append status %s should only have one partition %s"
+          .format(responseStatus, groupMetadataPartition))
+
+      // construct the commit response status and insert
+      // the offset and metadata to cache if the append status has no error
+      val status = responseStatus(groupMetadataPartition)
+
+      if (status.error != ErrorMapping.NoError) {
+        debug("Metadata from group %s with generation %d failed when appending to log due to %s"
+          .format(group.groupId, group.generationId, ErrorMapping.exceptionNameFor(status.error)))
+
+        // transform the log append error code to the corresponding the commit status error code
+        if (status.error == ErrorMapping.UnknownTopicOrPartitionCode)
+          ErrorMapping.ConsumerCoordinatorNotAvailableCode
+        else if (status.error == ErrorMapping.NotLeaderForPartitionCode)
+          ErrorMapping.NotCoordinatorForConsumerCode
+        else if (status.error == ErrorMapping.MessageSizeTooLargeCode
+          || status.error == ErrorMapping.MessageSetSizeTooLargeCode
+          || status.error == ErrorMapping.InvalidFetchSizeCode)
+          Errors.INVALID_COMMIT_OFFSET_SIZE.code
+        else
+          status.error
+      }
+
+      // notify the storing thread to return
+      responseLatch.countDown()
+    }
+
+    // call replica manager to append the offset messages
+    replicaManager.appendMessages(
+      config.offsetCommitTimeoutMs.toLong,
+      config.offsetCommitRequiredAcks,
+      true, // allow appending to internal offset topic
+      groupMetadataMessageSet,
+      putCacheCallback)
+
+    responseLatch.await()
```

Comment:
Ack. Fixing it now, the originally motivation is that with the callback handling with parking callbacks the unit test ```GroupCoordinatorResponseTest``` becomes extremely tricky.